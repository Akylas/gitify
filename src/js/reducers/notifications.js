import { Map, List, fromJS } from 'immutable';

import {
  NOTIFICATIONS_REQUEST, NOTIFICATIONS_SUCCESS, NOTIFICATIONS_FAILURE,
  MARK_NOTIFICATION_SUCCESS, MARK_REPO_NOTIFICATION_SUCCESS
} from '../actions';

const initialState = Map({
  response: List(),
  isFetching: false,
  failed: false
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case NOTIFICATIONS_REQUEST:
      return state
        .set('isFetching', true)
        .set('failed', false);
    case NOTIFICATIONS_SUCCESS:
      return state
        .set('isFetching', false)
        .set('response', fromJS(action.payload));
    case NOTIFICATIONS_FAILURE:
      return state
        .set('isFetching', false)
        .set('failed', true)
        .set('response', List());
    case MARK_NOTIFICATION_SUCCESS:
      return state
        .set('response', state.get('response')
          .filterNot((obj) => obj.get('id') === action.meta.id));
    case MARK_REPO_NOTIFICATION_SUCCESS:
      return state
        .set('response', state.get('response')
          .filterNot((obj) => obj.getIn(['repository', 'full_name']) === action.meta.repoSlug));
    default:
      return state;
  }
};
