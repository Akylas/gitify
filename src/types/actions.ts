export const TOGGLE_SETTINGS_MODAL = 'TOGGLE_SETTINGS_MODAL';

export interface ToggleSettingsModalAction {
  type: typeof TOGGLE_SETTINGS_MODAL;
}
