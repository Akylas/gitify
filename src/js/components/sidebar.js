import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { List, Map } from 'immutable';
import { shell } from 'electron';

import { fetchNotifications, logout, toggleSettingsModal } from '../actions';
import { isUserEitherLoggedIn } from '../utils/helpers';
import Logo from './logo';
import Constants from '../utils/constants';

export class Sidebar extends React.Component {
  componentDidMount() {
    const self = this;
    const iFrequency = 60000;

    this.requestInterval = setInterval(() => {
      self.refreshNotifications();
    }, iFrequency);
  }

  componentWillUnmount() {
    clearInterval(this.requestInterval);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.connectedAccounts > this.props.connectedAccounts) {
      this.props.fetchNotifications();
    }
  }

  refreshNotifications() {
    if (this.props.isEitherLoggedIn) {
      this.props.fetchNotifications();
    }
  }

  openBrowser() {
    shell.openExternal(`https://www.github.com/${Constants.REPO_SLUG}`);
  }

  _renderGitHubAccount() {
    const defaultHostname = Constants.DEFAULT_AUTH_OPTIONS.hostname;
    const notificationsCount = this.props.notifications
      .find(obj => obj.get('hostname') === defaultHostname, null, Map())
      .get('notifications', List()).size;

    return (
      <div className="badge-account" title={defaultHostname}>
        <div className="mr-auto name">GitHub</div>
        <div>{notificationsCount === 0 ? <span className="octicon octicon-check" /> : notificationsCount}</div>
      </div>
    );
  }

  _renderEnterpriseAccounts() {
    return this.props.enterpriseAccounts.map((account, idx) => {
      const splittedHostname = account.get('hostname').split('.');
      const accountDomain = splittedHostname[splittedHostname.length - 2];
      const notificationsCount = this.props.notifications
        .find(obj => obj.get('hostname') === account.get('hostname'), null, Map())
        .get('notifications', List()).size;

      return (
        <div
          key={idx}
          title={account.get('hostname')}
          className="badge-account"
          >
          <div className="mr-auto name">{accountDomain}</div>
          <div>{notificationsCount === 0 ? <span className="octicon octicon-check" /> : notificationsCount}</div>
        </div>
      );
    });
  }

  render() {
    const { hasStarred, isEitherLoggedIn, isGitHubLoggedIn, notifications } = this.props;
    const notificationsCount = notifications.reduce((memo, acc) => memo + acc.get('notifications').size, 0);

    return (
      <div className="sidebar-wrapper">
        <Logo onClick={this.openBrowser} />

        {isEitherLoggedIn && (
          <div className="badge badge-count text-success my-1">
            {notifications.isEmpty() ? 'All Read' : `${notificationsCount} Unread`}
          </div>
        )}

        {isEitherLoggedIn && (
          <ul className="nav nav-inline mb-2">
            <li className="nav-item text-white">
              <i
                title="Refresh"
                className="nav-link fa fa-refresh"
                onClick={() => this.refreshNotifications()} />
            </li>

            <li className="nav-item text-white">
              <i
                title="Settings"
                className="nav-link fa fa-cog"
                onClick={() => this.props.toggleSettingsModal()} />
            </li>
          </ul>
        )}

        {isGitHubLoggedIn && this._renderGitHubAccount()}
        {this._renderEnterpriseAccounts()}

        <div className="footer">
          {!!isEitherLoggedIn &&
            <Link
              to="/enterpriselogin"
              className="btn btn-block btn-sm btn-outline-secondary btn-add">Add <br />Enterprise</Link>
          }

          {!hasStarred && (
            <button className="btn btn-block btn-sm btn-outline-secondary btn-star" onClick={this.openBrowser}>
              <i className="fa fa-github" /> Star
            </button>
          )}
        </div>
      </div>
    );
  }
};

export function mapStateToProps(state) {
  const enterpriseAccounts = state.auth.get('enterpriseAccounts');
  const isGitHubLoggedIn = state.auth.get('token') !== null;
  const connectedAccounts = enterpriseAccounts.reduce((memo, acc) => memo + 1, isGitHubLoggedIn ? 1 : 0);

  return {
    isGitHubLoggedIn,
    isEitherLoggedIn: isUserEitherLoggedIn(state.auth),
    notifications: state.notifications.get('response'),
    enterpriseAccounts,
    connectedAccounts,
    hasStarred: state.settings.get('hasStarred'),
  };
};

export default connect(mapStateToProps, { fetchNotifications, logout, toggleSettingsModal })(Sidebar);
