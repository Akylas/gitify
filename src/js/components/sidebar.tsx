import * as React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { List, Map } from 'immutable';
import { shell } from 'electron';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCog, faSyncAlt } from '@fortawesome/free-solid-svg-icons';
import { faGithub } from '@fortawesome/free-brands-svg-icons';

import { fetchNotifications, logout, toggleSettingsModal } from '../actions';
import { isUserEitherLoggedIn } from '../utils/helpers';
import { LogoWhite } from './logos/white';
import Constants from '../utils/constants';
import { ToggleSettingsModalAction } from '../../types/actions';

interface IProps {
  fetchNotifications: () => void;
  connectedAccounts: number;

  enterpriseAccounts: any; // PropTypes.object.isRequired;
  notifications: any; // PropTypes.object.isRequired;

  toggleSettingsModal: () => ToggleSettingsModalAction;
  hasStarred: boolean;
  isEitherLoggedIn: boolean;
  isGitHubLoggedIn: boolean;
}

export class Sidebar extends React.Component<IProps> {
  requestInterval: NodeJS.Timer;

  componentDidMount() {
    const self = this;
    const iFrequency = 5000;

    this.requestInterval = setInterval(() => {
      self.refreshNotifications();
    }, iFrequency);
  }

  state = {
    connectedAccounts: [],
  };

  static getDerivedStateFromProps(props, state) {
    if (props.connectedAccounts > state.connectedAccounts) {
      props.fetchNotifications();
    }

    return {
      connectedAccounts: props.connectedAccounts,
    };
  }

  componentWillUnmount() {
    clearInterval(this.requestInterval);
  }

  refreshNotifications() {
    if (this.props.isEitherLoggedIn) {
      this.props.fetchNotifications();
    }
  }

  onOpenBrowser() {
    shell.openExternal(`https://www.github.com/${Constants.REPO_SLUG}`);
  }

  _renderGitHubAccount() {
    const { enterpriseAccounts, notifications } = this.props;
    const defaultHostname = Constants.DEFAULT_AUTH_OPTIONS.hostname;
    const notificationsCount = notifications
      .find(obj => obj.get('hostname') === defaultHostname, null, Map())
      .get('notifications', List()).size;

    return (
      <div
        className={`badge-account ${enterpriseAccounts.isEmpty() && 'last'}`}
        title={defaultHostname}
      >
        <div className="mr-auto name">GitHub</div>
        <div>
          {notificationsCount === 0 ? (
            <span className="octicon octicon-check" />
          ) : (
            notificationsCount
          )}
        </div>
      </div>
    );
  }

  _renderEnterpriseAccounts() {
    const { enterpriseAccounts } = this.props;

    return this.props.enterpriseAccounts.map((account, idx) => {
      const splittedHostname = account.get('hostname').split('.');
      const accountDomain = splittedHostname[splittedHostname.length - 2];
      const notificationsCount = this.props.notifications
        .find(
          obj => obj.get('hostname') === account.get('hostname'),
          null,
          Map()
        )
        .get('notifications', List()).size;

      return (
        <div
          key={idx}
          title={account.get('hostname')}
          className={`badge-account${
            enterpriseAccounts.size === idx + 1 ? ' last' : ''
          }`}
        >
          <div className="mr-auto name">{accountDomain}</div>
          <div>
            {notificationsCount === 0 ? (
              <span className="octicon octicon-check" />
            ) : (
              notificationsCount
            )}
          </div>
        </div>
      );
    });
  }

  render() {
    const {
      hasStarred,
      isEitherLoggedIn,
      isGitHubLoggedIn,
      notifications,
    } = this.props;
    const notificationsCount = notifications.reduce(
      (memo, acc) => memo + acc.get('notifications').size,
      0
    );

    return (
      <div className="sidebar-wrapper">
        <LogoWhite onClick={this.onOpenBrowser} />

        {isEitherLoggedIn && (
          <div className="badge badge-count text-success my-1">
            {notifications.isEmpty()
              ? 'All Read'
              : `${notificationsCount} Unread`}
          </div>
        )}

        {isEitherLoggedIn && (
          <ul className="nav nav-inline mb-2">
            <li className="nav-item text-white">
              <FontAwesomeIcon
                className="mx-1"
                icon={faSyncAlt}
                onClick={() => this.refreshNotifications()}
                title="Refresh"
                fixedWidth
              />
            </li>

            <li className="nav-item text-white">
              <FontAwesomeIcon
                className="mx-1"
                icon={faCog}
                onClick={() => this.props.toggleSettingsModal()}
                title="Settings"
                fixedWidth
              />
            </li>
          </ul>
        )}

        {isGitHubLoggedIn &&
          !this.props.enterpriseAccounts.isEmpty() &&
          this._renderGitHubAccount()}
        {this._renderEnterpriseAccounts()}

        <div className="footer">
          {!!isEitherLoggedIn && (
            <Link
              to="/enterpriselogin"
              className="btn btn-block btn-sm btn-outline-secondary btn-add"
            >
              Add <br />
              Enterprise
            </Link>
          )}

          {!hasStarred && (
            <button
              className="btn btn-block btn-sm btn-outline-secondary btn-star"
              onClick={this.onOpenBrowser}
            >
              <FontAwesomeIcon icon={faGithub} title="GitHub" /> Star
            </button>
          )}
        </div>
      </div>
    );
  }
}

export function mapStateToProps(state) {
  const enterpriseAccounts = state.auth.get('enterpriseAccounts');
  const isGitHubLoggedIn = state.auth.get('token') !== null;
  const connectedAccounts = enterpriseAccounts.reduce(
    memo => memo + 1,
    isGitHubLoggedIn ? 1 : 0
  );

  return {
    isGitHubLoggedIn,
    isEitherLoggedIn: isUserEitherLoggedIn(state.auth),
    notifications: state.notifications.get('response'),
    enterpriseAccounts,
    connectedAccounts,
    hasStarred: state.settings.get('hasStarred'),
  };
}

export default connect(mapStateToProps, {
  fetchNotifications,
  logout,
  toggleSettingsModal,
})(Sidebar);
