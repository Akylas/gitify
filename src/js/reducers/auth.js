import { Map } from 'immutable';
import { LOGIN, LOGOUT } from '../actions';

const initialState = Map({
  response: Map(),
  token: null,
  isFetching: false,
  failed: false
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN.REQUEST:
      return state
        .set('isFetching', true)
        .set('failed', false)
        .set('response', Map())
        .set('token', null);
    case LOGIN.SUCCESS:
      return state
        .set('isFetching', false)
        .set('token', action.payload.access_token);
    case LOGIN.FAILURE:
      return state
        .set('isFetching', false)
        .set('failed', true)
        .set('response', Map(action.payload));
    case LOGOUT:
      return state
        .set('response', null)
        .set('token', null);
    default:
      return state;
  }
};
