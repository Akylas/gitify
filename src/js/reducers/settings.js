import { UPDATE_SETTING, HAS_STARRED } from '../actions';

const initialState = {
  participating: false,
  playSound: true,
  showNotifications: true,
  markOnClick: false,
  openAtStartup: false,
  hasStarred: false
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case UPDATE_SETTING:
      return {
        ...state,
        [action.setting]: action.value
      };
    case HAS_STARRED.SUCCESS:
      return {
        ...state,
        hasStarred: true
      };
    case HAS_STARRED.FAILURE:
      return {
        ...state,
        hasStarred: false
      };
    default:
      return state;
  }
};
