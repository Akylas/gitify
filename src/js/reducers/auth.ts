import { LOGIN } from '../actions';
import { LOGOUT } from '../../types/actions';
import { AuthState, EnterpriseAccount } from '../../types/reducers';

const initialState: AuthState = {
  response: {},
  token: null,
  isFetching: false,
  failed: false,
  enterpriseAccounts: [],
};

export default function reducer(state = initialState, action): AuthState {
  switch (action.type) {
    case LOGIN.REQUEST:
      return {
        ...state,
        isFetching: true,
        failed: false,
        response: {},
      };
    case LOGIN.SUCCESS:
      if (action.isEnterprise) {
        const enterpriseAccount: EnterpriseAccount = {
          hostname: action.hostname,
          token: action.payload.access_token,
        };
        return {
          ...state,
          isFetching: false,
          enterpriseAccounts: [...state.enterpriseAccounts, enterpriseAccount],
        };
      }

      return {
        ...state,
        isFetching: false,
        token: action.payload.access_token,
      };
    case LOGIN.FAILURE:
      return {
        ...state,
        isFetching: false,
        failed: true,
        response: action.payload,
      };
    case LOGOUT:
      return {
        ...state,
        response: null,
        token: null,
        enterpriseAccounts: [],
      };
    default:
      return state;
  }
}
