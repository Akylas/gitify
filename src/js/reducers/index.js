import { combineReducers } from 'redux';
import * as storage from 'redux-storage';
import merger from 'redux-storage-merger-immutablejs';

import auth from './auth';
import notifications from './notifications';
import settings from './settings';

export default storage.reducer(
  combineReducers({
    auth,
    notifications,
    settings,
  }),
  merger
);
