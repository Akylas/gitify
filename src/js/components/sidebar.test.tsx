import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

const { shell, ipcRenderer } = require('electron');

import { Sidebar, mapStateToProps } from './sidebar';
import {
  mockedEnterpriseAccounts,
  mockedNotificationsRecuderData,
} from '../__mocks__/mockedData';
import { AuthState, AppState, SettingsState } from '../../types/reducers';

describe('components/Sidebar.tsx', () => {
  let clock;

  const props = {
    isFetching: false,
    isGitHubLoggedIn: true,
    isEitherLoggedIn: true,
    connectedAccounts: 2,
    enterpriseAccounts: mockedEnterpriseAccounts,
    notifications: mockedNotificationsRecuderData,
    notificationsCount: 4,
    hasStarred: false,
    fetchNotifications: jest.fn(),
    history: {
      goBack: jest.fn(),
      push: jest.fn(),
    },
    location: {
      pathname: '/',
    },
  };

  beforeEach(() => {
    clock = jest.useFakeTimers();

    spyOn(ipcRenderer, 'send');
    spyOn(shell, 'openExternal');
    spyOn(window, 'clearInterval');

    props.fetchNotifications.mockReset();
    props.history.push.mockReset();
  });

  afterEach(() => {
    clock.clearAllTimers();
  });

  it('should test the mapStateToProps method', () => {
    const state = {
      auth: {
        token: '12345',
        enterpriseAccounts: mockedEnterpriseAccounts,
      } as AuthState,
      notifications: {
        response: [],
      },
    } as AppState;

    const mappedProps = mapStateToProps(state);

    expect(mappedProps.isGitHubLoggedIn).toBeTruthy();
    expect(mappedProps.isEitherLoggedIn).toBeTruthy();
    expect(mappedProps.notifications).toBeDefined();
  });

  it('should render itself & its children (logged in)', () => {
    const tree = TestRenderer.create(
      <MemoryRouter>
        <Sidebar {...props} />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render itself & its children (logged out)', function() {
    const caseProps = {
      ...props,
      notifications: [],
      enterpriseAccounts: [],
      isGitHubLoggedIn: false,
      isEitherLoggedIn: false,
    };
    const tree = TestRenderer.create(
      <MemoryRouter>
        <Sidebar {...caseProps} />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should clear the interval when unmounting', () => {
    spyOn(Sidebar.prototype, 'componentDidMount').and.callThrough();

    const { unmount } = render(
      <MemoryRouter>
        <Sidebar {...props} />
      </MemoryRouter>
    );

    expect(Sidebar.prototype.componentDidMount).toHaveBeenCalledTimes(1);
    unmount();
    expect(window.clearInterval).toHaveBeenCalledTimes(1);
  });

  it('should load notifications after 60000ms', function() {
    const {} = render(
      <MemoryRouter>
        <Sidebar {...props} />
      </MemoryRouter>
    );

    expect(props.fetchNotifications).toHaveBeenCalledTimes(1);
    clock.runTimersToTime(60000);
    expect(props.fetchNotifications).toHaveBeenCalledTimes(2);
    clock.runTimersToTime(60000);
    expect(props.fetchNotifications).toHaveBeenCalledTimes(3);
  });

  it('should fetch the notifications if another account logs in', () => {
    const { rerender } = render(
      <MemoryRouter>
        <Sidebar {...props} />
      </MemoryRouter>
    );

    props.fetchNotifications.mockReset();

    rerender(
      <MemoryRouter>
        <Sidebar {...props} connectedAccounts={props.connectedAccounts + 1} />
      </MemoryRouter>
    );

    expect(props.fetchNotifications).toHaveBeenCalledTimes(1);
  });

  it('should refresh the notifications', () => {
    const { getByTitle } = render(
      <MemoryRouter>
        <Sidebar {...props} />
      </MemoryRouter>
    );
    props.fetchNotifications.mockReset();
    fireEvent.click(getByTitle('Refresh'));
    expect(props.fetchNotifications).toHaveBeenCalledTimes(1);
  });

  it('should go to the enterprise login route', () => {
    const { debug, getByLabelText } = render(
      <MemoryRouter>
        <Sidebar {...props} />
      </MemoryRouter>
    );
    fireEvent.click(getByLabelText('Login with GitHub Enterprise'));
    expect(props.history.push).toHaveBeenCalledTimes(1);
  });

  it('open the gitify repo in browser', () => {
    const { debug, getByLabelText } = render(
      <MemoryRouter>
        <Sidebar {...props} />
      </MemoryRouter>
    );
    fireEvent.click(getByLabelText('Star on GitHub'));
    expect(shell.openExternal).toHaveBeenCalledTimes(1);
  });
});
