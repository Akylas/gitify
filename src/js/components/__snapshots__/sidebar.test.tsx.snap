// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Sidebar.tsx should render itself & its children (logged in) 1`] = `
<div
  className="sc-bdVaJa VpBUF"
>
  <div
    className="sc-bwzfXH hudJrY"
  >
    <svg
      className="logo"
      onClick={[Function]}
      role="logo"
      viewBox="0 0 180.31 182.52"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <defs>
        <linearGradient
          gradientUnits="userSpaceOnUse"
          id="linear-white-gradient"
          x1="300.53"
          x2="130.22"
          y1="294.89"
          y2="240.53"
        >
          <stop
            offset="0"
            stopColor="#CCCCCC"
          />
          <stop
            offset="1"
            stopColor="#FFFFFF"
          />
        </linearGradient>
        <linearGradient
          id="linear-white-gradient-2"
          x1="315.88"
          x2="216.87"
          xlinkHref="#linear-white-gradient"
          y1="246.16"
          y2="214.57"
        />
        <linearGradient
          id="linear-white-gradient-3"
          x1="264.39"
          x2="147.04"
          xlinkHref="#linear-white-gradient"
          y1="211.56"
          y2="174.1"
        />
      </defs>
      <path
        d="M178.67,226.51a47.7,47.7,0,0,0,89.93,21.91h46.56a91.61,91.61,0,0,1-88.82,69.35c-50.45,0-91.49-40.94-91.49-91.26C145,186.6,171,205,178.67,226.51Z"
        style={
          Object {
            "fill": "url(#linear-white-gradient)",
          }
        }
        transform="translate(-134.85 -135.25)"
      />
      <path
        d="M214.16,226.51a21.91,21.91,0,0,1,21.91-21.91h38.57a12.08,12.08,0,0,1,8.3,4l32.21,39.86H236.07A21.91,21.91,0,0,1,214.16,226.51Z"
        style={
          Object {
            "fill": "url(#linear-white-gradient-2)",
          }
        }
        transform="translate(-134.85 -135.25)"
      />
      <path
        d="M258.74,179.07H226.34a47.61,47.61,0,0,0-47.66,47.44c-11.27,24.3-30.12,41.63-43.83,0,0-50.32,41-91.26,91.49-91.26h32.41a21.91,21.91,0,0,1,0,43.83Z"
        style={
          Object {
            "fill": "url(#linear-white-gradient-3)",
          }
        }
        transform="translate(-134.85 -135.25)"
      />
    </svg>
    <div
      className="sc-htpNat cKBiPV"
    >
      4 Unread
    </div>
    <div
      className="sc-bxivhb gbbeNv"
    >
      <div
        onClick={[Function]}
        title="Refresh"
      >
        <svg
          aria-hidden="true"
          className="sc-ifAKCX bkDHmA"
          height={18}
          role="img"
          style={
            Object {
              "display": "inline-block",
              "fill": "currentColor",
              "userSelect": "none",
              "verticalAlign": "text-bottom",
            }
          }
          viewBox="0 0 12 16"
          width={13.5}
        >
          <path
            d="M10.24 7.4a4.15 4.15 0 01-1.2 3.6 4.346 4.346 0 01-5.41.54L4.8 10.4.5 9.8l.6 4.2 1.31-1.26c2.36 1.74 5.7 1.57 7.84-.54a5.876 5.876 0 001.74-4.46l-1.75-.34zM2.96 5a4.346 4.346 0 015.41-.54L7.2 5.6l4.3.6-.6-4.2-1.31 1.26c-2.36-1.74-5.7-1.57-7.85.54C.5 5.03-.06 6.65.01 8.26l1.75.35A4.17 4.17 0 012.96 5z"
            fillRule="evenodd"
          />
        </svg>
      </div>
      <a
        href="/settings"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="sc-ifAKCX bkDHmA"
          height={18}
          role="img"
          style={
            Object {
              "display": "inline-block",
              "fill": "currentColor",
              "userSelect": "none",
              "verticalAlign": "text-bottom",
            }
          }
          viewBox="0 0 14 16"
          width={15.75}
        >
          <path
            d="M14 8.77v-1.6l-1.94-.64-.45-1.09.88-1.84-1.13-1.13-1.81.91-1.09-.45-.69-1.92h-1.6l-.63 1.94-1.11.45-1.84-.88-1.13 1.13.91 1.81-.45 1.09L0 7.23v1.59l1.94.64.45 1.09-.88 1.84 1.13 1.13 1.81-.91 1.09.45.69 1.92h1.59l.63-1.94 1.11-.45 1.84.88 1.13-1.13-.92-1.81.47-1.09L14 8.75v.02zM7 11c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
            fillRule="evenodd"
          />
        </svg>
      </a>
    </div>
    <div
      className="badge-account false"
      title="github.com"
    >
      <div
        className="mr-auto name"
      >
        GitHub
      </div>
      <div>
        2
      </div>
    </div>
    <div
      className="badge-account last"
      title="github.gitify.io"
    >
      <div
        className="mr-auto name"
      >
        gitify
      </div>
      <div>
        2
      </div>
    </div>
  </div>
  <div
    className="sc-EHOje kvIlCf"
  >
    <button
      aria-label="Login with GitHub Enterprise"
      className="sc-bZQynM eNBUPr"
      onClick={[Function]}
    >
      Add Enterprise
    </button>
    <button
      aria-label="Star on GitHub"
      className="sc-bZQynM eNBUPr"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="octicon"
        height={12}
        role="img"
        style={
          Object {
            "display": "inline-block",
            "fill": "currentColor",
            "userSelect": "none",
            "verticalAlign": "text-bottom",
          }
        }
        viewBox="0 0 16 16"
        width={12}
      >
        <path
          d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0016 8c0-4.42-3.58-8-8-8z"
          fillRule="evenodd"
        />
      </svg>
       Star
    </button>
  </div>
</div>
`;

exports[`components/Sidebar.tsx should render itself & its children (logged out) 1`] = `
<div
  className="sc-bdVaJa VpBUF"
>
  <div
    className="sc-bwzfXH hudJrY"
  >
    <svg
      className="logo"
      onClick={[Function]}
      role="logo"
      viewBox="0 0 180.31 182.52"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <defs>
        <linearGradient
          gradientUnits="userSpaceOnUse"
          id="linear-white-gradient"
          x1="300.53"
          x2="130.22"
          y1="294.89"
          y2="240.53"
        >
          <stop
            offset="0"
            stopColor="#CCCCCC"
          />
          <stop
            offset="1"
            stopColor="#FFFFFF"
          />
        </linearGradient>
        <linearGradient
          id="linear-white-gradient-2"
          x1="315.88"
          x2="216.87"
          xlinkHref="#linear-white-gradient"
          y1="246.16"
          y2="214.57"
        />
        <linearGradient
          id="linear-white-gradient-3"
          x1="264.39"
          x2="147.04"
          xlinkHref="#linear-white-gradient"
          y1="211.56"
          y2="174.1"
        />
      </defs>
      <path
        d="M178.67,226.51a47.7,47.7,0,0,0,89.93,21.91h46.56a91.61,91.61,0,0,1-88.82,69.35c-50.45,0-91.49-40.94-91.49-91.26C145,186.6,171,205,178.67,226.51Z"
        style={
          Object {
            "fill": "url(#linear-white-gradient)",
          }
        }
        transform="translate(-134.85 -135.25)"
      />
      <path
        d="M214.16,226.51a21.91,21.91,0,0,1,21.91-21.91h38.57a12.08,12.08,0,0,1,8.3,4l32.21,39.86H236.07A21.91,21.91,0,0,1,214.16,226.51Z"
        style={
          Object {
            "fill": "url(#linear-white-gradient-2)",
          }
        }
        transform="translate(-134.85 -135.25)"
      />
      <path
        d="M258.74,179.07H226.34a47.61,47.61,0,0,0-47.66,47.44c-11.27,24.3-30.12,41.63-43.83,0,0-50.32,41-91.26,91.49-91.26h32.41a21.91,21.91,0,0,1,0,43.83Z"
        style={
          Object {
            "fill": "url(#linear-white-gradient-3)",
          }
        }
        transform="translate(-134.85 -135.25)"
      />
    </svg>
  </div>
  <div
    className="sc-EHOje kvIlCf"
  >
    <button
      aria-label="Star on GitHub"
      className="sc-bZQynM eNBUPr"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="octicon"
        height={12}
        role="img"
        style={
          Object {
            "display": "inline-block",
            "fill": "currentColor",
            "userSelect": "none",
            "verticalAlign": "text-bottom",
          }
        }
        viewBox="0 0 16 16"
        width={12}
      >
        <path
          d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0016 8c0-4.42-3.58-8-8-8z"
          fillRule="evenodd"
        />
      </svg>
       Star
    </button>
  </div>
</div>
`;
